src/py21cmmc/likelihood.py

```diff
@@ -930,42 +930,117 @@ class LikelihoodLuminosityFunction(LikelihoodBaseFile):
+                "can only pass a single datafile to LikelihoodLuminosityFunction!"
+            )
+        if self.noisefile is not None and len(self.noisefile) != 1:
+            raise ValueError(
+                "can only pass a single noisefile to LikelihoodLuminosityFunction!"
+            )
+
+        self.name = str(name)
+
+    def setup(self):
+        super().setup()
+
+        # We only allow one datafile, so get the data out of it
+        # to make it easier to work with.
+        self.data = self.data[0]
+        if self.noise is not None:
+            self.noise = self.noise[0]
+
+    def _read_data(self):
+        data = super()._read_data()
+
+        # data only has one entry (only one file allowed)
+        # if that entry has values with one dimension, add a first
+        # dimension as a redshift dimension of size 1.
+        for key, val in data[0].items():
+            if not hasattr(val[0], "__len__"):
+                data[0][key] = np.atleast_2d(val)
+
+        return data
+
+    def _read_noise(self):
+        noise = super()._read_noise()
+
+        # data only has one entry (only one file allowed)
+        # if that entry has values with one dimension, add a first
+        # dimension as a redshift dimension of size 1.
+        for key, val in noise[0].items():
+            if not hasattr(val[0], "__len__"):
+                noise[0][key] = np.atleast_2d(val)
+
+        return noise
+
+    @cached_property
+    def paired_core(self):
+        """The luminosity function core that is paired with this likelihood."""
+        paired = []
+        for c in self._cores:
+            if isinstance(c, core.CoreLuminosityFunction) and c.name == self.name:
+                paired.append(c)
+        if len(paired) > 1:
+            raise ValueError(
+                "You've got more than one CoreLuminosityFunction with the same name -- they will overwrite each other!"
+            )
+        return paired[0]
+
     @property
     def redshifts(self):
-        return self.core_primary.redshift
+        return self.paired_core.redshift
 
     def reduce_data(self, ctx):
-        return ctx.get("luminosity_function")
+        lfunc = ctx.get("luminosity_function" + self.name)
+        if not self._is_setup:
+            # During setup, return a list, so that it can be matched with the
+            # list of length one of datafile to be written.
+            return [lfunc]
+        else:
+            return lfunc
 
     def computeLikelihood(self, model):
         lnl = 0
         for i, z in enumerate(self.redshifts):
             model_spline = InterpolatedUnivariateSpline(
                 model["Muv"][i][::-1], model["lfunc"][i][::-1]
             )
-
             lnl += -0.5 * np.sum(
-                (self.data["lfunc"][i] - model_spline(self.data["Muv"][i])) ** 2
-                / self.noise[i]["sigma"] ** 2
+                (self.data["lfunc"][i] - 10 ** model_spline(self.data["Muv"][i])) ** 2
```

Comment:
I'm not familiar with the subtleties of the LF likelihood, so another pair of eyes should check that this is correct.

Reply:
Don't worry about this -- it was merged to this branch by accident by @binnietom. It will be reviewed in its own PR (#32 )